### DISCUSSION TITLE
question: Float formatting

### DISCUSSION THREAD

**Discussion by xhpohanka** [taocpp/json#68] [Q&A]
Hello,
can I somehow control formatting of floats in resulting json string? Basically I would like to limit the maximal count of digits...

Similar code
```
float val = 40.9;
string key = "Humidity";
tao::json::value v = { { key, val } };
auto str = tao::json::to_string(v);
cout << str << endl;
```
results in
`{"Humidity":40.900001525878906}`

best regards
Jan

**Comment by d-frey** [comment]
Do you actually need to control the number of digits or are you looking for the shortest (but correct) representation?

@ColinH Would an update of the conversion libraries help to improve the current overlong representations?

**Comment by xhpohanka** [comment]
I would prefer something like %0.2f in printf :D 
Seriously - we use JSON also for human readable output from our sensor that definitely has lesser accuracy than double/float .

**Comment by ColinH** [comment]
@xhpohanka This is not currently possible, but should be easy to add given the modularity of the library.
@d-frey I'm wondering why 40.9 produces such a long output, we need to check the bit-pattern whether it's one of the innocent-looking but complicated-to-encode-in-base-2 numbers, in other word we (also) need to check whether our current output library is doing the right thing (in general).

**Comment by ColinH** [comment]
The number `40.9` does not have a direct representation as `double` as it seems to translate to an  infinitely repeating sequence in binary (just like 1/3 doesn't have a finite decimal representation). I believe that the library we use to format floating point numbers does not choose to format the `double` that `40.9` converts to as `40.9` because while it would be correct in the sense that it is a shorter decimal number that converts to the same bit pattern, it does not faithfully represent the binary number as precisely as possible...

**Comment by ColinH** [comment]
I've added an example that shows how to use `printf()` for doubles, does that work for now?
